schematic/base.py

```
@@ -0,0 +1,346 @@
+        """
+        for i, col in enumerate(self.columns):
+            if col.name == column.name:
+                self.columns[i] = column
+                return
+        raise ValueError(
+            "No such column name {} in {}".format(
+                column.name, self.name))
+
+    def get_rows(self, *args, **kwargs):
+        """Generator for rows of the table described by this TableDefinition.
+
+        Args:
+          *args, **kwargs: DB-specific arguments for connection, e.g.,
+                           a psycopg2.connection or csv.DictReader object
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+
+class Schematic(DictableMixin, object):
+    """Interface for implementation specifics for a type of database or warehouse.
+
+    The TableColumnTypes in a given schematic form a tree with the most
+    restrictive types being leaf nodes and the least restrictive type
+    being the root node.
+
+    Attributes:
+      name: Static attribute with the name of this schematic
+      most_restrictive_types: the leaf nodes of the restrictivity tree
+      table_def: implementation of TableDefinition for this schematic
+    """
+    name = 'schematic'
+    most_restrictive_types = []
+    table_definition_class = TableDefinition
+
+    def get_distance_from_leaf_node(self, column_type):
+        """Get the distance between the givenx TableColumnType
+        and its nearest leaf node.
+
+        Returns:
+          An int
+        """
+        distances = []
+        for ct in self.most_restrictive_types:
+            distance = 0
+            nlr = ct
+            while nlr:
+                if nlr == column_type:
+                    distances.append(distance)
+                    nlr = None
+                else:
+                    distance += 1
+                    nlr = nlr.next_less_restrictive
+        return min(distances)
```

In the event that column_type isn't in the tree anywhere, let's return a special error

---

Good call, updated.